[WIP] JIT Static Hooks: cpp tests

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49547 [WIP] JIT Static Hooks: cpp tests**
* #49546 [WIP] JIT Static Hooks: python tests
* #49545 [WIP] JIT Static Hooks: serialization logic
* #49975 [WIP] JIT Static Hooks: schema checking logic
* #49544 [WIP] JIT Static Hooks: adding hooks to class type and adding logic for hook running/compilation

Adds tests to verify that hooks are importable and runnable from CPP with the intended behavior. 

Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

Commit:
[WIP] JIT Static Hooks: cpp tests

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Lilyjjo commented:
Wait, sorry, I didn't see how many errors appeared since I added the schema checking PR, hold off on reviewing

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"




[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

SplitInfinity commented on test/jit_hooks/model.py:
```diff
@@ -0,0 +1,515 @@
+import argparse
```
The classes and tests in this file look identical to those in the unit tests from the PR below this one. If that is the case, I think we should explore ways to put them in some common location and use them in the Python unit tests. The only difference appears to be that there are no error tests here (which makes sense) and these modules are exported instead of being passed to `checkModule`.

Lilyjjo replied:
I moved the modules and hooks to a new shared file 👍

SplitInfinity reviewed (commented):
Nice!

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

SplitInfinity commented on .jenkins/pytorch/macos-test.sh:
```diff
@@ -134,18 +134,40 @@ test_custom_script_ops() {
+  popd
+
+  # Run tests Python-side and export a script module.
+  # python test_jit_hooks.py -v
```
```suggestion
```

SplitInfinity commented on test/jit_hooks/model.py:
```diff
@@ -0,0 +1,113 @@
+    global save_name
+    save_name = options.export_script_module_to + "_"
+
+    test_submodule_forward_single_input()
+    test_submodule_forward_multiple_inputs()
+    test_submodule_multiple_hooks_single_input()
+    test_submodule_multiple_hooks_multiple_inputs()
+    test_submodule_hook_return_nothing()
+    test_submodule_same_hook_repeated()
+
+    test_module_forward_single_input()
+    test_module_forward_multiple_inputs()
+    test_module_multiple_hooks_single_input()
+    test_module_multiple_hooks_multiple_inputs()
+    test_module_hook_return_nothing()
+    test_module_same_hook_repeated()
+
+    test_module_no_forward_input()
+    test_forward_tuple_input()
```
It seems that there's a one-to-one map from test case -> function that creates the model for that test case, and that all of the functions above are the same. You can probably create just one function that takes in a filename and model and then call that in a loop over a list of `(test_name, test_model)`. Something like:
```
tests = [
  ("test_submodule_forward_single_input", test_submodule_forward_single_input()),
  ...
]

for name, model in tests:
  m_scripted = torch.jit.script(model)
  filename = ... # generated using name
  torch.jit.save(model, filename)
```

Lilyjjo replied:
😮 
that's some next level refactoring -- done ✅

SplitInfinity commented on test/jit_hooks/model.py:
```diff
@@ -0,0 +1,113 @@
+
+def test_submodule_hook_return_nothing():
+    m_scripted = torch.jit.script(test_submodule_hook_return_nothing_model())
+    m_scripted.save(save_name + "test_submodule_hook_return_nothing" + ".pt")
```
How come `save` is used here instead of `torch.jit.save`? Importing issues?

Lilyjjo replied:
no reason besides me not knowing the difference between the two

SplitInfinity commented on test/jit_hooks/test_jit_hooks.cpp:
```diff
@@ -0,0 +1,270 @@
+  std::cout << "testing: test_module_forward_invocation_no_hooks_run" << std::endl;
+  test_module_forward_invocation_no_hooks_run(path_to_exported_script_module);
+
+  std::cout << "JIT CPP Hooks okay!" << std::endl;
```
Nit: It's probably automatically inserted but we are using the return value in a shell script sooo
```suggestion
  std::cout << "JIT CPP Hooks okay!" << std::endl;
  return 0;
```

SplitInfinity commented on test/jit_hooks/test_jit_hooks.cpp:
```diff
@@ -0,0 +1,270 @@
+
+#include <iostream>
+
+void test_submodule_multiple_hooks_single_input(
```
I think there is opportunity for refactoring here as well - a lot of the tests look the same. You could aggregate model paths, inputs, and outputs into some list or something and then run the same test code (load, run, compare output) for all `(file, input, output)` tuples. I think you should be able to deal with varying input/output types by expressing them as `IValues` and relying on `IValue::equals` for comparison.

SplitInfinity commented on test/jit_hooks/test_jit_hooks.cpp:
```diff
@@ -0,0 +1,270 @@
+
+  auto output = module(inputs);
+  std::cout << "----- module output: " << output << std::endl;
+  std::ostringstream stream;
+  stream << output;
+  std::string output_str =  stream.str();
+  AT_ASSERT("([pre_hook_override_name, inner_mod_name], pre_hook_override2_fh1_fh2)" == output_str);
```
Could you use `IValue` comparison here instead by creating an `ivalue::Tuple` that contains the expected list and string?

SplitInfinity commented on test/jit_hooks/test_jit_hooks.cpp:
```diff
@@ -0,0 +1,270 @@
+
+  auto output = module(inputs);
+  std::cout << "----- module output: " << output << std::endl;
+  std::ostringstream stream;
+  stream << output;
+  std::string output_str =  stream.str();
+  AT_ASSERT("([pre_hook_override_name2, outer_mod_name, inner_mod_name], pre_hook_override_fh1_fh2)" == output_str);
```
Same here.

SplitInfinity commented on test/jit_hooks/test_jit_hooks.cpp:
```diff
@@ -0,0 +1,270 @@
+  inputs.push_back(input_tuple);
+
+  auto output = module(inputs);
+  std::cout << "----- module output: " << output << std::endl;
+  std::ostringstream stream;
+  stream << output;
+  std::string output_str =  stream.str();
+  AT_ASSERT("(11,)" == output_str);;
```
Same here.

SplitInfinity reviewed (commented):
I still think there is further opportunity for refactoring/code reuse but based on my track record with this stack of diffs so far I'm probably wrong. Let me know what you think!

SplitInfinity commented on test/jit_hooks/model.py:
```diff
@@ -0,0 +1,113 @@
+    test_module_multiple_hooks_multiple_inputs()
+    test_module_hook_return_nothing()
+    test_module_same_hook_repeated()
+
```
Also, how come there are no C++ tests for direct forward invocation not invoking hooks?

Lilyjjo replied:
It's included in the cpp testing code already as 'test_module_forward_invocation_no_hooks_run', it's not in the .py file because it can re-use another test's saved model

SplitInfinity replied:
Ah, I see it now. I would also add to that a similar test to the one I commented on in the Python test PR: one where you access a submodule, store it in a temporary variable, and call that using `Module::operator()`.

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Adds tests to verify that hooks are importable and runnable from CPP with the intended behavior. 

Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Adds tests to verify that hooks are importable and runnable from CPP with the intended behavior. 

Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Adds tests to verify that hooks are importable and runnable from CPP with the intended behavior. 

Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]

Commit:
Update on "[WIP] JIT Static Hooks: cpp tests"


Adds tests to verify that hooks are importable and runnable from CPP with the intended behavior. 

Differential Revision: [D25771118](https://our.internmc.facebook.com/intern/diff/D25771118)

[ghstack-poisoned]
